File: python/src/tracking_client.py
import types
from .tracking_service import TrackingService
from tcp_messages import MessageClient, Message
from cellworld import *


class TrackingClient (MessageClient):
    def __init__(self):
        MessageClient.__init__(self)
        self.registered_consumer = False
        self.on_step = None
        self.current_states = dict()
        self.router.add_route("_step$", self.__process_step__, Step)

    def __process_step__(self, step: Step):
        try:
            self.current_states[step.agent_name] = step.copy()
            if self.on_step:
                self.on_step(step)
        finally:
            pass

    def connect(self, ip: str = "127.0.0.1") -> bool:
        return MessageClient.connect(self, ip, TrackingService.port())

    def register_consumer(self, call_back=None):
        if call_back:
            check_types(call_back, [types.FunctionType, types.MethodType], "incorrect type for call_back")
        self.on_step = call_back
        response = self.send_request(Message("register_consumer"), 0)
        self.registered_consumer = (response.body == "success")
        return self.registered_consumer

    def unregister_consumer(self):
        response = self.send_request(Message("unregister_consumer"), 0)
        if response.body == "success":
            self.registered = False
            return True
        else:
            return False

    def get_current_state(self, agent_name: str) -> Step:
        return self.current_states[agent_name]

    def contains_agent_state(self, agent_name: str) -> bool:
        return agent_name in self.current_states

    def set_throughput(self, throughput) -> bool:
        return self.send_request(Message("set_throughput", throughput)).get_body(bool)

File: src/tracking_client.cpp
#include <agent_tracking/tracking_client.h>
#include <agent_tracking/tracking_service.h>
#include <stdexcept>

using namespace cell_world;
using namespace std;
using namespace agent_tracking;
using namespace tcp_messages;

namespace agent_tracking {

    bool Tracking_client::connect() {
        return connect("127.0.0.1");
    }

    bool Tracking_client::connect(const std::string &ip) {
        return connect(ip, Tracking_service::get_port());
    }

    bool Tracking_client::connect(const std::string &ip, int port) {
        if (local_server) return true;
        is_connected = Message_client::connect(ip, port);
        return is_connected;
    }

    bool Tracking_client::register_consumer() {
        return subscribe();
    }

    void Tracking_client::send_step(const cell_world::Step &step) {
        if (is_connected) {
            if (local_server) {
                local_server->send_step(step);
            } else {
                send_message(Message("send_step", step));
            }
        } else {
            this->on_step(step);
        }
    }

    bool Tracking_client::unregister_consumer() {
        return unsubscribe();
    }

    Tracking_client::Tracking_client() {

    }

    cell_world::Step &Tracking_client::get_current_state(const string &agent_name) {
        return current_states[agent_name];
    }

    void Tracking_client::on_step(const cell_world::Step &step) {
        current_states[step.agent_name] = step;
    }

    bool Tracking_client::contains_agent_state(const string &agent_name) {
        return current_states.count(agent_name) != 0;
    }

    Tracking_client::~Tracking_client() {
        easy_tcp::Client::disconnect();
    }

    bool Tracking_client::subscribe() {
        if (local_server) {
            return local_server->subscribe_local(this);

        } else {
            return Message_client::subscribe();
        }
    }

    bool Tracking_client::unsubscribe() {
        if (local_server) {
            return local_server->unsubscribe_local(this);

        } else {
            return Message_client::unsubscribe();
        }
    }
}
